File: 1168-duplicate-zeros/1168-duplicate-zeros.py
class Solution:
    def duplicateZeros(self, arr: List[int]) -> None:
        """
        Do not return anything, modify arr in-place instead.
        """
        idx = 0
        n = len(arr)
        while idx < n:            
            if arr[idx] == 0 and idx != n-1:
                stop = idx + 1
                itr = n-1
                while stop < itr:
                    arr[itr] = arr[itr-1]
                    itr -= 1
                arr[stop] = 0
                idx += 1
            idx += 1

        

File: 1168-duplicate-zeros/1168-duplicate-zeros.java
class Solution {
    public void duplicateZeros(int[] arr) {
        //2 passs
        int possibleDuplicate = 0;
        int len = arr.length-1;
        for(int index = 0; index <= len - possibleDuplicate; index++){
            if(arr[index] == 0){
                if(index == len - possibleDuplicate){
                    arr[len] = 0;
                    len -= 1;
                    break;
                }
                possibleDuplicate++;
            }
        }

        int last = len - possibleDuplicate;
        for(int index = last; index >= 0; index-- ){
            if(arr[index] == 0){
                arr[index + possibleDuplicate] = 0;
                possibleDuplicate--;
                arr[index + possibleDuplicate] = 0;
            }else{
                arr[index + possibleDuplicate] = arr[index];
            }
        }
    }
}